server/schedule/operator/operator.go

```diff
@@ -658,40 +658,59 @@ func CreateTransferLeaderOperator(desc string, region *core.RegionInfo, sourceSt
 
 // CreateMoveRegionOperator creates an operator that moves a region to specified stores.
 func CreateMoveRegionOperator(desc string, cluster Cluster, region *core.RegionInfo, kind OpKind, storeIDs map[uint64]struct{}) (*Operator, error) {
+	mvkind, steps, err := moveRegionSteps(region, storeIDs, false, cluster)
+	if err != nil {
+		return nil, err
+	}
+	kind |= mvkind
+	return NewOperator(desc, region.GetID(), region.GetRegionEpoch(), kind, steps...), nil
+}
+
+func moveRegionSteps(region *core.RegionInfo, storeIDs map[uint64]struct{}, force bool, cluster Cluster) (OpKind, []OpStep, error) {
 	var steps []OpStep
+	var kind OpKind
+	oldStores := make([]uint64, 0, len(storeIDs))
+	newStores := make([]uint64, 0, len(storeIDs))
 	// Add missing peers.
 	for id := range storeIDs {
 		if region.GetStorePeer(id) != nil {
+			oldStores = append(oldStores, id)
 			continue
 		}
+		newStores = append(newStores, id)
 		peer, err := cluster.AllocPeer(id)
 		if err != nil {
-			return nil, err
-		}
-		if cluster.IsRaftLearnerEnabled() {
-			steps = append(steps,
-				AddLearner{ToStore: id, PeerID: peer.Id},
-				PromoteLearner{ToStore: id, PeerID: peer.Id},
-			)
-		} else {
-			steps = append(steps, AddPeer{ToStore: id, PeerID: peer.Id})
+			log.Debug("peer alloc failed", zap.Error(err))
+			return kind, nil, err
 		}
+		steps = append(steps, CreateAddPeerSteps(id, peer.Id, cluster)...)
+		kind |= OpRegion
 	}
 
+	// Transferring leader to a new follower may be refused by TiKV if the
+	//  apply process is slow (the new follower haven't been promoted).
+	// So, the new leader should be an old follower if it is possible,
+	//  otherwise the first suitable new follower.
+	storeIDsVec := append(oldStores, newStores...)
+
 	// Remove redundant peers.
 	for _, peer := range region.GetPeers() {
 		if _, ok := storeIDs[peer.GetStoreId()]; ok {
 			continue
 		}
 		if region.GetLeader().GetId() == peer.GetId() {
-			for targetID := range storeIDs {
-				steps = append(steps, TransferLeader{FromStore: peer.GetStoreId(), ToStore: targetID})
-				break
+			tlkind, tlsteps := transferLeaderToAnySteps(peer.GetStoreId(), storeIDsVec, force, cluster)
```

Comment:
And the before, we transfer leader to the first added peer, now seems not satisfied now.